### PR TITLE
Multi

### PR DESCRIPTION
**Pull request by maweigert**
This PR contains a lot of new features

### Support for multi class prediction 

StarDist models now optionally includes a multiclass head that outputs dense multiclass probabilities. 

This is activated by setting the `n_classes`  argument of the `config` object to a positive number, signifying the number of different foreground classes. The resulting multiclass probability image will have `1+n_classes` channels with the first channel being the background probability. A backward compatible model without such a head can be obtained by setting `n_classes = None` (which is the default).

``` python
n_classes = None      # no multiclass prediction (old style model, default)
n_classes = 1         # predicts background/foreground
n_classes = 3         # predicts background + 3 different foreground classes
 
conf = Config2D(...,  n_classes = n_classes, ...)
```
If `n_classes = None` or `n_classes=1` the model training can be done as usual: 

```python
model.train(X, Y, validation_data = (X_v,Y_v),...) 
```
If `n_classes>1` one has to provide a sequence that contains for each training mask a dictionary that maps each label id of the mask to a class id  `(0,...,n_classes)` with `0` being background.

```python
classes = ({1:1,2:1,3:2,4:2,5:2} , {8:1,9:1,10:2} , ... )
model.train(X, Y, classes = classes, validation_data = (X_v,Y_v,classes_v),...) 
```

Afterwards, the result dictionary of `model.predict_instances` contains for each found instance the predicted probability for each class:
 
```python
n_classes = 4

...

label, res = model.predict_instances(img)

prob_class = res["prob_class"]

print(prob_class.shape)

>>> (101,5)

```

### Improved NMS efficiency with sparse prediction and KDtree neighborhood selection 

Prediction of large images should now be way faster (2-3x) and use way less (2-5x) memory (comparable to `predict_instances_big`)

* To reduce memory pressure one can activate sparse prediction (retaining only valid points per tile) by using 

```python 
model.predict_instances(img,... , sparse = True)
```
* Both 2D and 3D models now use the same coordinate/distance layout 

* Fast neighborhood selection with KDTrees via [nanoflann](https://github.com/jlblancoc/nanoflann) 

 
### Other Changes

* Allow multithreaded data generation via 
```python
model.train(..., workers=n_workers, ...)
```

### CODE REVIEW

File: stardist/__init__.py [hunk 1]
```diff
@@ -7,8 +7,15 @@ def format_Warning(message, category, filename, lineno, line=''):
 
 from .version import __version__
 
-# TODO: which functions to expose here? all?
+import warnings
+from pathlib import Path
+def warning_on_one_line(message, category, filename, lineno, file=None, line=None):
+    # return f'%s:%s: %s:%s\n'.format(filename, lineno, category.__name__, message)
+    return f"({Path(filename).name}) {category.__name__}: {message}\n"
+warnings.formatwarning = warning_on_one_line
```
Comment (uschmidt83) [inline stardist/__init__.py:15]: Why?

File: stardist/geometry/geom2d.py [hunk 2]
```diff
@@ -97,40 +120,84 @@ def polygons_to_label(coord, prob, points, shape=None, thr=-np.inf):
...
+    points.shape = (n_polys, 2)
+    return coord.shape = (n_polys,2,n_rays)
+    """
+    dist = np.asarray(dist)
+    points = np.asarray(points)
+    assert dist.ndim==2 and points.ndim==2 and len(dist)==len(points) and points.shape[1]==2
+    n_rays = dist.shape[1]
+    phis = ray_angles(n_rays)
+    coord = points[...,np.newaxis] + (dist[:,np.newaxis]*np.array([np.sin(phis),np.cos(phis)]))
+    return coord
```
Comment (uschmidt83) [inline stardist/geometry/geom2d.py:135]: Redefinition of `dist_to_coord` breaks `other2D` notebooks, need to be updated.

File: stardist/geometry/geom2d.py [hunk 3]
```diff
@@ -97,40 +120,84 @@ def polygons_to_label(coord, prob, points, shape=None, thr=-np.inf):
...
+    coord = np.asarray(coord)
+    if labels is None: labels=np.arange(1,len(coord)+1)
+
+    assert coord.ndim==3 and coord.shape[1]==2  and len(coord)==len(labels)
+
+    lbl = np.zeros(shape,np.int32)
+    
+    for i,c in zip(labels,coord):
+        rr,cc = polygon(*c, shape)
+        lbl[rr,cc] = i+1
```
Comment (uschmidt83) [inline stardist/geometry/geom2d.py:152]: `i` or `i+1`?

File: stardist/__init__.py (hunk 1 above)
Comment (maweigert) [reply stardist/__init__.py]: Remnant of an old branch. Fixed

File: stardist/geometry/geom2d.py (hunk 3 above)
Comment (maweigert) [reply stardist/geometry/geom2d.py:152]: Fixed

Comment (maweigert) [reply stardist/geometry/geom2d.py]: Fixed

**Comment by uschmidt83** [comment]
- Duplicate code due to `sparse` and "regular" variants

**Comment by uschmidt83** [comment]
Maybe need to speed up some tests?

```
============================== slowest 20 durations ==============================
22.45s call     tests/test_model2D.py::test_model_multiclass[1-auto-3]
22.31s call     tests/test_model2D.py::test_model_multiclass[3-classes2-3]
20.12s call     tests/test_model2D.py::test_model_multiclass[None-auto-1]
18.34s call     tests/test_big.py::test_predict3D
13.62s call     tests/test_model3D.py::test_model[33-grid1-1-resnet-4-False]
13.17s call     tests/test_model2D.py::test_speed
7.86s call     tests/test_model3D.py::test_model[73-grid0-None-resnet-1-False]
6.09s call     tests/test_model3D.py::test_model[7-grid2-2-unet-1-True]
5.43s call     tests/test_model3D.py::test_optimize_thresholds
3.05s call     tests/test_model3D.py::test_predict_dense_sparse
3.02s call     tests/test_big.py::test_predict2D[False]
2.70s call     tests/test_model2D.py::test_pretrained_scales
2.51s call     tests/test_model2D.py::test_model[17-grid0-None-1-False]
2.30s call     tests/test_big.py::test_predict2D[True]
1.95s call     tests/test_model3D.py::test_model_multiclass[1-auto-3]
1.86s call     tests/test_model2D.py::test_model[4-grid2-2-1-True]
1.77s call     tests/test_model2D.py::test_model[32-grid1-1-4-False]
1.76s call     tests/test_model3D.py::test_model_multiclass[3-classes2-3]
1.52s call     tests/test_nms2D.py::test_large
1.47s call     tests/test_model3D.py::test_model_multiclass[None-auto-1]
```

**Review by uschmidt83 — CHANGES_REQUESTED** [review]

**Comment by maweigert** [comment]
Todo: 

[ ] Add sparse prediction modality to 3D

File: stardist/lib/stardist3d_impl.cpp [hunk 4]
```diff
@@ -930,15 +957,16 @@ void _COMMON_non_maximum_suppression_sparse(
                     const float* scores, const float* dist, const float* points,
                     const int n_polys, const int n_rays, const int n_faces, 
                     const float* verts, const int* faces,
-                    const float threshold, const int use_bbox, const int verbose, 
+                    const float threshold, const int use_bbox, const int use_kdtree,
```
[the remark this answers is not in the document]
Comment (maweigert) [reply stardist/lib/stardist3d_impl.cpp:960]: Fixed

File: stardist/nms.py [hunk 5]
```diff
@@ -58,19 +74,183 @@ def non_maximum_suppression(coord, prob, grid=(1,1), b=2, nms_thresh=0.5, prob_t
...
+
+    if verbose:
+        print("keeping %s/%s polygons" % (np.count_nonzero(inds), len(inds)))
+        print("NMS took %.4f s" % (time() - t))
+
+    return points[inds], scores[inds], dist[inds]
+
+
+def non_maximum_suppression_sparse(dist, prob, points, b=2, nms_thresh=0.5,
+                                   use_bbox=True, use_kdtree = True, verbose=False):
```
[the remark this answers is not in the document]
Comment (maweigert) [reply stardist/nms.py:136]: Fixed

File: stardist/lib/stardist2d.cpp [hunk 6]
```diff
@@ -352,10 +385,242 @@ static PyObject* c_non_max_suppression_inds (PyObject *self, PyObject *args) {
...
 
+  
+static PyObject* c_non_max_suppression_inds(PyObject *self, PyObject *args) {
+
+  PyArrayObject *dist=NULL,*points_arr=NULL, *mapping=NULL, *result=NULL;
+  float threshold;
+  int verbose, use_kdtree, use_bbox;
+
+  if (!PyArg_ParseTuple(args, "O!O!iiif", &PyArray_Type, &dist, &PyArray_Type, &points_arr ,
+                        &use_kdtree, &use_bbox, &verbose, &threshold))
```
[the remark this answers is not in the document]
Comment (maweigert) [reply stardist/lib/stardist2d.cpp:396]: Fixed

File: stardist/models/base.py [hunk 7]
```diff
@@ -70,20 +71,79 @@ def kld(y_true, y_pred):
     return K.mean(K.binary_crossentropy(y_true, y_pred) - K.binary_crossentropy(y_true, y_true), axis=-1)
 
 
+def masked_iou_loss(mask, reg_weight=0, norm_by_mask=True):
+    def iou_loss(y_true, y_pred):
+        axis = -1 if backend_channels_last() else 1 
+
+        inter = K.mean(K.sign(y_pred)*K.square(K.minimum(y_true,y_pred)), axis = axis)
```
Comment (uschmidt83) [inline stardist/models/base.py:78]: K.sign?

File: stardist/geometry/geom2d.py [hunk 8]
```diff
@@ -97,40 +120,84 @@ def polygons_to_label(coord, prob, points, shape=None, thr=-np.inf):
...
+    assert len(points)==len(prob) and points.shape[1]==2 and prob.ndim==1 
+
+    n_rays = dist.shape[1]
+    
+    ind = prob>thr
+    points = points[ind]
+    dist = dist[ind]
+    prob = prob[ind]
+
+    ind = np.argsort(prob)
```
[the remark this answers is not in the document]
Comment (maweigert) [reply stardist/geometry/geom2d.py]: Yes, using the stable version should indeed be better. Thx!

File: stardist/models/base.py (hunk 7 above)
Comment (maweigert) [reply stardist/models/base.py:78]: Yes, for negative y_pred (distances) iou should be negative - otherwise one could get arbitrary high iou by simply setting y_pred to very large negative numbers

File: stardist/models/base.py [hunk 9]
```diff
@@ -381,6 +650,9 @@ def predict_instances(self, img, axes=None, normalizer=None,
         normalizer : :class:`csbdeep.data.Normalizer` or None
             (Optional) normalization of input image before prediction.
             Note that the default (``None``) assumes ``img`` to be already normalized.
+        sparse: bool 
+            If true, aggregate probabilities/distances sparsely during tiled 
+            prediction to save memory (recommended). Currently only 3D supported
```
Comment (uschmidt83) [inline stardist/models/base.py:614]: Only 3D supported?

File: stardist/models/model2d.py [hunk 10]
```diff
@@ -291,8 +343,12 @@ def train(self, X, Y, validation_data, augmenter=None, seed=None, epochs=None, s
...
-            Tuple of X,Y validation arrays.
+        classes (optional): "auto" or iterable of same length as X 
+             label -> class mapping for each mask if multiclass prediction is activated (n_classes>0)
+             list of dicts with label -> class id (1,...,n_classes) 
+             "auto" -> all objects will be assigned to the first foreground class 
+        validation_data : tuple(:class:`numpy.ndarray`, :class:`numpy.ndarray`) or triple (if multiclass)
+            Tuple (triple if multiclass) of X,Y validation arrays.
         augmenter : None or callable
             Function with expected signature ``xt, yt = augmenter(x, y)``
             that takes in a single pair of input/label image (x,y) and returns
```
Comment (uschmidt83) [inline stardist/models/model2d.py:354]: Should augmenter be able to change class ids?

File: stardist/models/model2d.py [hunk 11]
```diff
@@ -323,9 +379,17 @@ def simple_augmenter(x,y):
         if steps_per_epoch is None:
             steps_per_epoch = self.config.train_steps_per_epoch
 
+        if self._is_multiclass() and classes is None:
+            warnings.warn("Ignoring given classes as n_classes is set to None")
```
Comment (uschmidt83) [inline stardist/models/model2d.py:383]: Don't understand  (`self._is_multiclass()` is True => n_classes is NOT set to None)

Comment (uschmidt83) [reply stardist/models/model2d.py:383]: What does classes set to None even mean?

File: stardist/models/model2d.py [hunk 12]
```diff
@@ -65,21 +71,32 @@ def __getitem__(self, i):
...
 
-        return [X], [prob,dist]
+        if self.n_classes is None:
+            return [X], [prob,dist]
+        else:
+            prob_class = np.stack(tuple((mask_to_categorical(y, self.n_classes, self.classes[k]) for y,k in zip(Y, idx))))
+
+            # as it prob_class will be later upscaled, usign zoom here leads to better registered maps
+            # prob_class = prob_class[self.ss_grid]
+            prob_class = zoom(prob_class, tuple(1/s for s in self.ss_grid_factor)+(1,), order=0)
```
Comment (uschmidt83) [inline stardist/models/model2d.py:97]: Is this still true?

File: stardist/models/model3d.py [hunk 13]
```diff
@@ -67,30 +75,36 @@ def __getitem__(self, i):
         if X.ndim == 4: # input image has no channel axis
             X = np.expand_dims(X,-1)
 
-        tmp = [edt_prob(lbl, anisotropy=self.anisotropy) for lbl in Y]
+        tmp = np.stack([edt_prob(lbl, anisotropy=self.anisotropy) for lbl in Y])
```
Comment (uschmidt83) [inline stardist/models/model3d.py:78]: I think the reason for `self.out_edt_prob` and `self.out_star_dist3D` was that `np.stack` was the time bottleneck. This advantage is now gone...

File: stardist/models/model3d.py [hunk 14]
```diff
@@ -67,30 +75,36 @@ def __getitem__(self, i):
...
-        return [X], [prob,dist]
 
+        if self.n_classes is None:
+            return [X], [prob,dist]
+        else:
+            prob_class = np.stack(tuple((mask_to_categorical(y, self.n_classes, self.classes[k]) for y,k in zip(Y, idx))))
+
+            # as it prob_class will be later upscaled, usign zoom here leads to better registered maps
+            # prob_class = prob_class[self.ss_grid]
+            prob_class = zoom(prob_class, tuple(1/s for s in self.ss_grid_factor)+(1,), order=0)
```
Comment (uschmidt83) [inline stardist/models/model3d.py:103]: Again, still relevant/true?

File: stardist/models/model3d.py [hunk 15]
```diff
@@ -174,7 +192,7 @@ class Config3D(BaseConfig):
         .. _ReduceLROnPlateau: https://keras.io/callbacks/#reducelronplateau
     """
 
-    def __init__(self, axes='ZYX', rays=None, n_channel_in=1, grid=(1,1,1), anisotropy=None, backbone='resnet', **kwargs):
+    def __init__(self, axes='ZYX', rays=None, n_channel_in=1, grid=(1,1,1), n_classes = None, anisotropy=None, backbone='unet', **kwargs):
```
Comment (uschmidt83) [inline stardist/models/model3d.py:195]: Default backbone `unet` now?

File: stardist/models/model3d.py [hunk 16]
```diff
@@ -456,52 +538,117 @@ def simple_augmenter(x,y):
...
+            # map the overlap_label to something positive and back
+            # (as relabel_sequential doesn't like negative values)
+            if overlap_label is not None and overlap_label<0 and (overlap_label in labels):
+                overlap_mask = (labels == overlap_label)
+                overlap_label2 = max(set(np.unique(labels))-{overlap_label})+1
+                labels[overlap_mask] = overlap_label2
+                labels, fwd, bwd = relabel_sequential(labels)
+                labels[labels == fwd[overlap_label2]] = overlap_label
+            else:
+                labels, _,_ = relabel_sequential(labels)
```
Comment (uschmidt83) [inline stardist/models/model3d.py:621]: Why necessary?

File: stardist/models/model2d.py (hunk 12 above)
Comment (uschmidt83) [reply stardist/models/model2d.py]: Change comment

File: stardist/models/model3d.py (hunk 13 above)
Comment (maweigert) [reply stardist/models/model3d.py]: Should be fixed

**Review by uschmidt83 — APPROVED** [review]
